Better ranked search results in app list

pierlo-upitup opened the pull request:
This fixes mesosphere/marathon#3018

ACs

- First result should be the exact app name match.
- Followed by deepest group match first.

From:

![screen shot 2016-01-19 at 12 29 32](https://cloud.githubusercontent.com/assets/1078545/12424240/84279814-becf-11e5-993f-7620dd232a97.png)

To:

<img width="1421" alt="screen shot 2016-01-19 at 16 57 25" src="https://cloud.githubusercontent.com/assets/1078545/12424244/875e192c-becf-11e5-97f8-6348a0004a13.png">

aldipower commented on src/js/components/AppListComponent.jsx:
```diff
@@ -186,7 +198,11 @@ var AppListComponent = React.createClass({
       nodesSequence = nodesSequence
-        .filter(app => app.id.indexOf(filterText) !== -1);
+        .filter(app => {
+          app.filterScore = score(app.id, filterText);
```
You are modifying the ```app``` here, this should be done only in the store (with emiting an event afterwards).
Also this is done in a ```filter```. It's not expectable that a filter modifies data.
So this isn't the right approach. Sorry. :)

Let's try to sort and score the list where the sorting belongs to.
Could you add a sort function after the .sort()-function at line 316?
```javascript
      if (filterText != null) {
        nodesSequences.sort((a, b) => {
          return score(a.id, filterText) > score(b.id, filterText)
            ? -1
            : 1;
        });
      }
```
Didn't tested it, but it should go into this direction.
Maybe there is also a more efficient way by integrating this into the already existing sort-function. Not sure.

This doesn't modify the app itself. :)

pierlo-upitup replied:
You're totally right. Will follow your suggestions.

aldipower replied:
:)

pierlo-upitup replied:
That won't work, however, if we plan on supporting manual overriding of the sorting. I would like, for example, to search for `phil` and get the ranked sorting. But if I then click on one of the sorting columns, (for example CPU), this should override the search rank. 
My solution is hacky so it's a no-go, so let's have a chat about a better solution that addresses all of that.

aldipower replied:
add a simple condition to the `if` should do the trick:
```
if (filterText != null && sortKey !== "id") {
```

pierlo-upitup replied:
That means i can't sort my search results alphabetically?

aldipower replied:
Alphabetically sorting is done before. 
https://github.com/mesosphere/marathon-ui/blob/master/src/js/components/AppListComponent.jsx#L315

The sort-function mentioned above should be added below that.

I did a mistake in the condition, must be:
```
if (filterText != null && sortKey === "id") {
```

aldipower replied:
Sorry, the links goes to the wrong view type I think. May bad.
But doesn't change the principle how it works.

An additional sort function should be added here: 

https://github.com/mesosphere/marathon-ui/blob/master/src/js/components/AppListComponent.jsx#L301

aldipower commented:
A test would be neat! :)

aldipower commented:
According to the branchname, this is actually a feature and not a fix.
But this is very very minor nitpicking.

aldipower commented:
I had a look at the fuzzaldrin code.
It looks slightly inefficient and it's not because it is coffescript. :)
A lot of replaces and matches there.
https://github.com/atom/fuzzaldrin/tree/master/src
I wonder if this works together without our scalability objective, when we will have a lot of apps.

Is there is maybe a simpler solution.

Maybe just taking a second sorting round only on the applications name without the groups-portion?
Then we had 2 sort steps (like with fuzzaldrin anyways).
1. Sorting on the complete id -> this affects mostly the sorting on the groups segements (like it is already)
2. Sorting only at the application name -> kind of a ranking

Sounds simple. Do I miss something?

pierlo-upitup commented:
I picked `fuzzAldrin` specifically because it adds fuzzy matching: 

> This library is used by Atom and so its focus will be on scoring and filtering paths, methods, and other things common when writing code. It therefore will specialize in handling common patterns in these types of strings such as characters like /, -, and _, and also handling of camel cased text.

That is a very useful feature to have in our toolbelt, imho. Imagine searching very nested appIds, without the need to be precise. 

Regarding performance, I will do some tests. Surely `indexOf` is more performant, but this does not look bad at all:

```
Filtering 66672 entries for 'index' took 146ms for 6168 results
Matching 66672 entries for 'index' took 140ms for 6168 results
```
(I ran `npm run benchmark` in the fuzzaldrin source code repo).

I agree we should add some tests, and I'll do that next.

pierlo-upitup commented:
@aldipower I've squashed the changes, please take another look now. 

Here's a gist that helps you generate a number of fake apps for testing performance and ranking: https://gist.github.com/pierlo-upitup/c325692b3e9511da6590

pierlo-upitup commented:
Should we disable the sorting caret for the `Name` column when `filterText != null`?

aldipower commented:
Thanxman! Going to take a second round now. 
Having problems to install the modules on the first try. I think the shrinkwrap is faulty.

pierlo-upitup commented:
Oh yeah. I need to re-generate it...

aldipower commented on src/js/components/AppListComponent.jsx:
```diff
@@ -186,7 +187,10 @@ var AppListComponent = React.createClass({
       nodesSequence = nodesSequence
-        .filter(app => app.id.indexOf(filterText) !== -1);
+        .filter(app => {
+          return score(app.id, filterText) > 0.025;
```
Why is this needed? It leads to unexpected behavior.

Before (like it was), I see all result (ranked with the sorting later):
![score-before](https://cloud.githubusercontent.com/assets/859154/12449096/231fe1dc-bf7b-11e5-9024-315fd333a57b.png)
:+1: 

After - one result is hidden, I don't expect this:
![score-after](https://cloud.githubusercontent.com/assets/859154/12449106/38dbcdba-bf7b-11e5-9a44-3edfd13e5534.png)
:-1: 

Let's stick with indexOf() and keep it simple.

pierlo-upitup replied:
It's just a matter of tweaking that 0.25 to something lower, maybe 0.2.

aldipower replied:
But what sense? 
And what else is hidden then unexpected?
Cannot detect the use case for the scoring here.

pierlo-upitup replied:
The score value is our way of determining matches in the app list for the given filterText. https://github.com/atom/fuzzaldrin#scorestring-query 

If we move back to indexOf we lose the whole fuzz matching algorithm completely! Why would you want to do that? I agree that the result not showing up was a regression, but that's because i set the score threshold too high. Please try again now.

aldipower replied:
OK, now I got it. You want fuzzy matched results, besides direct matches.
This makes sense.
I was confused because this behavior isn't reflect in the related issue.
It only says that the given results should be _ranked_ better. That's how I understand it.

With this, now, you modify the complete result set.
But I like it! 

Confirmative. :)

pierlo-upitup replied:
:tada:

aldipower commented on src/js/components/AppListComponent.jsx:
```diff
@@ -305,6 +310,14 @@ var AppListComponent = React.createClass({
+      if (filterText != null && sortKey === "id") {
+        nodesSequence = nodesSequence.sort((a, b) => {
+          return score(a.id, filterText) > score(b.id, filterText)
+            ? -1
```
```javascript
            ? -1 * sortDirection
            : 1 * sortDirection;
```
Enables sorting for the `id`, sorry, lacked in my example I've given you.

aldipower commented on src/js/components/AppListComponent.jsx:
```diff
@@ -296,6 +300,7 @@ var AppListComponent = React.createClass({
       appsStatusesCount: getInitialFilterCounts(AppStatus),
       appsHealthCount: getInitialFilterCounts(HealthStatus)
     };
+    var filterText = props.filters[FilterTypes.TEXT];
```
```javascript
let filterText = props.filters[FilterTypes.TEXT];
```
inside the if-block below.
It's only used there.

pierlo-upitup replied:
good point!

pierlo-upitup replied:
wait. the filterText value is in the condition of the `if` which is checked anyway every time. why should i miss one "cache" hit?

pierlo-upitup replied:
nvm ! i got it wrong.

aldipower commented:
@pierlo-upitup OK! Review complete. :)
Some things need a little bit rework, but after that I am going to merge this.

I am fine by going with fuzzaldrin, he is doing a good job. :) And my solution I thought about is _too_ simple, doesn't really work out.

This makes the search result way meaningful.
Thanx for the work @pierlo-upitup !

pierlo-upitup commented:
Thanks for that! I just squashed your requested changes and especially tweaked the score value to 0.02 – that should be fine! Please let me know your thoughts.

FTR: I think at some point it might be a really good idea to move this whole filtering logic to the AppsStore or to a new ApssFilterStore because this component is doing far too much!

aldipower commented:
Thx, please look at my comment regarding the "0.02".
Don't get the use case for that.
Let's keep it simple please.

aldipower commented on CHANGELOG.md:
```diff
@@ -16,6 +16,7 @@
 - \#2831 - Convert alert, confirm and prompt dialogs to new design
 - \#2909 - Update dialog messages
 - \#2655 - Never preselect dangerous actions in modal dialogs
+- \#3018 - Better search result ranking
```
Could you be a little bit more precise here? That we do a fuzzy search.

aldipower commented:
Cool! 
So, maybe a little bit more precise changelog, a test and a rebase is left.
*hurray*

pierlo-upitup commented:
I have removed the caret from the Name column when a filterText is present, as I believe it only generates confusion (you don't order search results on Google...). I've also added some tests. We should be good now!

pierlo-upitup commented:
@aldipower please take a look at this, for a clarification as to why I think we should not allow Name sorting when showing ranked search results:
![weird](https://cloud.githubusercontent.com/assets/1078545/12452516/2b644928-bf8f-11e5-8305-3d290af66ca3.gif)

When i click on the "Name" column, as a user I expect the results sorting by Name. But what this accomplishes, is to sort results by rank in the opposite direction instead.

aldipower commented:
@pierlo-upitup Yeah, got what you mean, I would add a hover text on the caret, that explains the sorting by score.
I like to have this functionality of sorting the name column by score.
"Name" is the title of the column, not the definition of sorting the column by id alphabetically.

aldipower commented:
The health bar is also not alphabetically sorted, but ranked! Same thing.

pierlo-upitup commented:
The Name column works *exactly* on the id value itself *outside* of the search page. 
Why should that behaviour change in this context to work on a completely different and invisible value (score)? The tooltip explanation feels like a workaround to me, and changing the behaviour of a sorting column is bad UX.

Sorting by the healthbar is consistent across all other pages, so it's not a good comparison.

Why would a user you want to see the worst ranking results first, anyway?

pierlo-upitup commented:
Take a look at Google Drive.The same is true: sorting by Name is disabled in the search results page.

aldipower commented:
OK, the context change is a good argument to me. Thank you for argue this out. The healthbar example wasn't good.
Disabling the caret is also a workaround, but I think probably the better one.

aldipower commented on src/js/components/AppListComponent.jsx:
```diff
@@ -343,7 +356,10 @@ var AppListComponent = React.createClass({
+    var filterText = this.props.filters[FilterTypes.TEXT];
+
+    if (sortKey === this.state.sortKey &&
+      !(sortKey === "id" && filterText != null && filterText !== "")) {
```
Sorry for bothering you, could we negate this?
```JS
(sortKey !== "id" || filterText == null || filterText === "")) {
```

pierlo-upitup replied:
forcepushed, sir!

aldipower commented:
Thank you for the tests, very cool!

pierlo-upitup commented:
Aahaha and of course coveralls.io decided to die on us on the first PR with some tests :stuck_out_tongue:

aldipower commented:
LOL

aldipower commented:
Great!!!1!ELEVEN!! 
Thanx a lot for your patience on my comments. 

Very well done!

This will be merged now. 

![merger-ahead](https://cloud.githubusercontent.com/assets/859154/12454425/b55264fa-bf97-11e5-99e8-c09cdd2569f5.jpg)

aldipower commented:
Mission accomplished!
